src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,248 @@
+// It will cause dead-lock.
+void QuotaAggregatorImpl::SetFlushCallback(FlushCallback callback) {
+  InternalSetFlushCallback(callback);
+}
+
+// If the quota could not be handled by the cache, returns NOT_FOUND,
+// caller has to send the request to service control.
+// Otherwise, returns OK and cached response.
+::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* response) {
+  if (request.service_name() != service_name_) {
+    return Status(Code::INVALID_ARGUMENT,
+                  (string("Invalid service name: ") + request.service_name() +
+                   string(" Expecting: ") + service_name_));
+  }
+
+  if (!request.has_allocate_operation()) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "allocate operation field is required.");
+  }
+
+  if (!cache_) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+  if (!lookup.Found()) {
+    // Create a temporary element which will be remained in the cache until
+    // actual response arrives
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse temp_response;
+    CacheElem* cache_elem =
+        new CacheElem(temp_response, SimpleCycleTimer::Now());
+    cache_elem->set_signature(request_signature);
+    cache_->Insert(request_signature, cache_elem, 1);
+
+    // By returning NOT_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  lookup.value()->Aggregate(request);
+
+  *response = lookup.value()->quota_response();
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// Caches a response from a remote Service Controller AllocateQuota call.
+::google::protobuf::util::Status QuotaAggregatorImpl::CacheResponse(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response) {
+  if (!cache_) {
+    return ::google::protobuf::util::Status::OK;
+  }
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+
+  CacheElem* cache_elem = new CacheElem(response, SimpleCycleTimer::Now());
+  cache_elem->set_signature(request_signature);
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+  if (lookup.Found()) {
```

Comment:
The logic should be:
* if a positive response, drop it.  since a temp positive response already inserted when sending request.
* if a negative response,  insert it to the cache. replace the one in the cache.

I don't see that logic here.